inc/saiswitch.h

```diff
@@ -2256,6 +2297,44 @@ typedef enum _sai_switch_attr_t
      */
     SAI_SWITCH_ATTR_SUPPORTED_FAILOVER_MODE,
 
+    /**
+     * @brief Tunnel mappers for encap
+     *
+     * @type sai_object_list_t
+     * @flags CREATE_AND_SET
+     * @objects SAI_OBJECT_TYPE_TUNNEL_MAP
+     * @default empty
+     */
+    SAI_SWITCH_ATTR_TUNNEL_ENCAP_ECN_MAPPERS,
```

Comment:
The new API assumes one global configuration for all tunnel types. Meaning, if we have IP in IP tunnel and VXLAN tunnel, and we set this attribute, they will both use the same map.
Some ASICs have capability to set it per tunnel type. Meaning, having SAI_SWITCH_ATTR_TUNNEL_IPINIP_ENCAP_ECN_MAPPERS, SAI_SWITCH_ATTR_TUNNEL_VXLAN_ENCAP_ECN_MAPPERS,
SAI_SWITCH_ATTR_TUNNEL_MPLS_ECN_MAPPERS (similar to hash API where we have global configuration per traffic type SAI_SWITCH_ATTR_ECMP_HASH_IPV4, SAI_SWITCH_ATTR_ECMP_HASH_IPV4_IN_IPV4, etc.)

Reply:
Thats correct. Idea is to use global switch level configuration if there is a uniform ecn mode for given tunnel types for encap or decap. 
This can always be overridden by the tunnel specific ECN mapper.

Reply:
Still we don't allow global per tunnel type
There can be several ways to configure :
1) Global per all tunnels
2) Global per tunnel type (meaning, all IP in IP tunnels, all VXLAN tunnels, etc.)
3) Per specific tunnel (what we have today)
This PR is adding #1, but not adding #2 which is more flexible. I suggest adding #2 instead of #1

Reply:
Lets discuss this. Currently this PR is supporting (1)

Reply:
I think having per tunnel type level of default setting as suggested by @itaibaz is good which gives more granularity option to the application user.  Perhaps we can still have an attribute that covers for ALL tunnel types as well such as SAI_SWITCH_ATTR_ALL_TUNNEL_ECN_MAPPERS that sets the default at switch level when there are no specific per Tunnel type nor at the per tunnel level.

Reply:
Thanks @gechiang @itaibaz  for comments.
Looks like ECN mappers are easier to extend as it is an object list.
But ECN mode is of type sai_tunnel_decap_ecn_mode_t. So either we create these enums per tunnel type (sai_tunnel_decap/encp/_vxlan/ipinip/mpls_ecn_mode_t
or just leave the ecn mode to support model 1. What do you think?

Reply:
I would prefer per tunnel type attributes
Also, not sure I follow the usage of single ecn mappers attribute with object list. We can attach multiple maps to a single attrib, but how will we specify which map is valid for IP in IP and which one is valid for VXLAN ?

Reply:
@itaibaz can you confirm if this is what you are requesting. 
- sai_tunnel_decap_vxlan_ecn_mode_t
- sai_tunnel_encap_vxlan_ecn_mode_t
- sai_tunnel_decap_ipinip_ecn_mode_t
- sai_tunnel_encap_ipinip_ecn_mode_t
- sai_tunnel_decap_mpls_ecn_mode_t
- sai_tunnel_encap_mpls_ecn_mode_t

Reply:
@itaibaz It makes sense to keep it consistent with tunnel ecn mode and provide per tunnel type attributes.

Change Done !!

https://github.com/opencomputeproject/SAI/pull/1153/commits/7347e0b87b95f283d4e1d0311ffbc934a298de6c